scene/3d/collision_object_3d.cpp

```diff
@@ -207,26 +223,40 @@ void CollisionObject3D::_clear_debug_shapes() {
 		ShapeData::ShapeBase *shapes = shapedata.shapes.ptrw();
 		for (int i = 0; i < shapedata.shapes.size(); i++) {
 			ShapeData::ShapeBase &s = shapes[i];
-			if (s.debug_shape) {
-				s.debug_shape->queue_delete();
-				s.debug_shape = nullptr;
-				--debug_shape_count;
+			if (s.debug_shape.is_valid()) {
+				RS::get_singleton()->free(s.debug_shape);
+				s.debug_shape = RID();
+				while (s.shape.is_valid() && !s.shape->is_connected("changed", callable_mp(this, &CollisionObject3D::_update_shape_data))) {
+					s.shape->disconnect("changed", callable_mp(this, &CollisionObject3D::_update_shape_data));
+				}
 			}
 		}
 	}
-
-	debug_shape_count = 0;
+	debug_shapes_count = 0;
 }
 
 void CollisionObject3D::_update_shape_data(uint32_t p_owner) {
-	if (is_inside_tree() && get_tree()->is_debugging_collisions_hint() && !Engine::get_singleton()->is_editor_hint()) {
+	if (_are_collision_shapes_visible()) {
 		if (debug_shapes_to_update.is_empty()) {
-			call_deferred("_update_debug_shapes");
+			callable_mp(this, &CollisionObject3D::_update_debug_shapes).call_deferred({}, 0);
 		}
 		debug_shapes_to_update.insert(p_owner);
 	}
 }
 
+void CollisionObject3D::_on_transform_changed() {
+	if (debug_shapes_count > 0 && !debug_shape_old_transform.is_equal_approx(get_global_transform())) {
+		debug_shape_old_transform = get_global_transform();
+		for (Map<uint32_t, ShapeData>::Element *E = shapes.front(); E; E = E->next()) {
```

Comment:
Better use lowercase here

Reply:
There are several places where that E appears looking in that source (not introduced by me). I can obviously change it, but to be consistent any occurrence of it should be changed.

Reply:
@AndreaCatania Is this change important to you? I don't know if we have clear rules about naming iterators for containers, but it's true `E` for `Element` is used in many places in the engine at the moment.

Reply:
Yes `E` for `Element` is a common pattern all over the codebase whenever we use a `Map`.